File: 백준 2023 신기한 소수 -백트랙킹-/Main.java
import java.util.Scanner;

class Main {
    private static int N;
    private static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        dfs("", 0);
        System.out.println(sb.toString());
    }

    private static void dfs(String num, int depth) {
        if (depth == N) {
            sb.append(num).append("\n");
            return;
        }
        for (int i = 1; i <= 9; i++) {
            int currentNum = Integer.parseInt(num + i);
            if (isPrime(currentNum)) {
                dfs(currentNum + "", depth + 1);
            }
        }

    }

    private static boolean isPrime(int n) {
        // 0, 1
        if (n < 2) return false;
        // 2
        if (n == 2) return true;

        for (int i = 2; i < n; i++) {
            // 소수 X
            if (n % i == 0) return false;
        }
        return true;
    }


}

File: 백준 2023 신기한 소수 -백트랙킹-/Main.kt
import java.util.*


private var N = 0
private val sb = StringBuilder()
fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    N = sc.nextInt()
    dfs("", 0)
    println(sb.toString())
}

private fun dfs(num: String, depth: Int) {
    if (depth == N) {
        sb.append(num).append("\n")
        return
    }
    for (i in 1..9) {
        val currentNum = (num + i).toInt()
        if (isPrime(currentNum)) {
            dfs(currentNum.toString() + "", depth + 1)
        }
    }
}

private fun isPrime(n: Int): Boolean { // 0, 1
    if (n < 2) return false
    // 2
    if (n == 2) return true
    for (i in 2 until n) { // 소수 X
        if (n % i == 0) return false
    }
    return true
}